### PR TITLE
Refactor connectivity into new components

### PR DESCRIPTION
**Pull request by remonh87**

### CODE REVIEW

**Comment by remonh87** [comment]
Got rid of the prescan connector and increased test coverage. Also got rid of a nasty cycle

File: lib/src/device_connector.dart [hunk 1]
```diff
@@ -40,4 +69,147 @@ class DeviceConnector {
 
     return autoconnectingRepeater.stream;
   }
+
+  Stream<ConnectionStateUpdate> connectToAdvertisingDevice({
```
Comment (remonh87) [inline lib/src/device_connector.dart:73]: From here it is just copy past the logic from the prescanconnector

File: lib/src/device_connector.dart [hunk 2]
```diff
@@ -1,28 +1,57 @@
...
     @required PluginController pluginController,
-  })  : assert(connectionStateUpdateStream != null),
-        assert(pluginController != null),
-        _connectionStateUpdateStream = connectionStateUpdateStream,
-        _controller = pluginController;
+    @required ScanSession Function() getCurrentScan,
+    @required DiscoveredDevicesRegistry discoveredDevicesRegistry,
+    @required
+        Stream<DiscoveredDevice> Function(
+                {List<Uuid> withServices, ScanMode scanMode})
```
Comment (werediver) [inline lib/src/device_connector.dart:21]: I'd add a trailing comma.

File: lib/src/model/generic_failure.dart [hunk 3]
```diff
@@ -13,4 +13,18 @@ class GenericFailure<T> {
 
   @override
   String toString() => "$runtimeType(code: $code, message: \"$message\")";
+
+  @override
+  bool operator ==(dynamic other) =>
```
Comment (werediver) [inline lib/src/model/generic_failure.dart:18]: I'd prefer not to use `dynamic` at all (but `Object`).

File: lib/src/reactive_ble.dart [hunk 4]
```diff
@@ -111,24 +106,17 @@ class FlutterReactiveBle {
...
-    );
+    _pluginController ??= PluginControllerFactory(_methodChannel).create();
 
     _initialization ??= _methodChannel.invokeMethod("initialize");
     _deviceConnector ??= DeviceConnector(
-      connectionStateUpdateStream: connectedDeviceStream,
       pluginController: _pluginController,
-    );
-
-    _prescanConnector = PrescanConnector(
```
Comment (werediver) [inline lib/src/reactive_ble.dart:125]: A-a-and, it's gone.

File: lib/src/model/generic_failure.dart (hunk 3 above)
Comment (remonh87) [reply lib/src/model/generic_failure.dart:18]: fixed

File: lib/src/converter/protobuf_converter.dart [hunk 5]
```diff
@@ -46,22 +46,24 @@ class ProtobufConverter {
...
-        failure: genericFailureFrom(
-          hasFailure: deviceInfo.hasFailure(),
-          getFailure: () => deviceInfo.failure,
-          codes: ConnectionError.values,
-          fallback: (rawOrNull) =>
-              rawOrNull == null ? null : ConnectionError.unknown,
-        ),
-      );
+  ConnectionStateUpdate connectionStateUpdateFrom(List<int> data) {
+    final deviceInfo = pb.DeviceInfo.fromBuffer(data);
```
[the remark this answers is not in the document]
Comment (remonh87) [reply lib/src/converter/protobuf_converter.dart:50]: yes for testing the plugincontroller and the facade things becomes much easier with this convention. I am going to change this when I refactor the rest of the methods into the new structure.

**Pull request merged**